core/src/main/java/com/linecorp/armeria/server/ExceptionReportingServerErrorHandler.java

```diff
@@ -0,0 +1,165 @@
+import java.util.concurrent.atomic.LongAdder;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.linecorp.armeria.common.AggregatedHttpResponse;
+import com.linecorp.armeria.common.HttpResponse;
+import com.linecorp.armeria.common.HttpStatus;
+import com.linecorp.armeria.common.RequestHeaders;
+import com.linecorp.armeria.common.annotation.Nullable;
+import com.linecorp.armeria.common.metric.MeterIdPrefix;
+import com.linecorp.armeria.common.util.TextFormatter;
+import com.linecorp.armeria.internal.common.metric.MicrometerUtil;
+import com.linecorp.armeria.server.logging.LoggingService;
+
+import io.micrometer.core.instrument.MeterRegistry;
+
+/**
+ * A {@link ServerErrorHandler} that wraps another {@link ServerErrorHandler}
+ * to periodically report the exceptions that were not logged by decorators such as
+ * {@link LoggingService}.
+ */
+final class ExceptionReportingServerErrorHandler implements ServerErrorHandler, ServerListener {
+
+    private static final Logger logger = LoggerFactory.getLogger(ExceptionReportingServerErrorHandler.class);
+    private static final AtomicIntegerFieldUpdater<ExceptionReportingServerErrorHandler> scheduledUpdater =
+            AtomicIntegerFieldUpdater.newUpdater(ExceptionReportingServerErrorHandler.class,
+                                                 "scheduled");
+
+    private final Duration interval;
+    private final ServerErrorHandler delegate;
+    private final MicrometerCounter micrometerCounter;
+    private final LongAdder counter;
+    private long lastExceptionsCount;
+    private volatile int scheduled;
+    @Nullable
+    private Throwable thrownException;
+    @Nullable
+    private ScheduledFuture<?> reportingTaskFuture;
+
+    ExceptionReportingServerErrorHandler(MeterRegistry meterRegistry, ServerErrorHandler serverErrorHandler,
+                                         Duration interval) {
+        assert !interval.isNegative() && !interval.isZero();
+        this.interval = interval;
+        delegate = serverErrorHandler;
+        micrometerCounter =
+                MicrometerUtil.register(meterRegistry,
+                                        new MeterIdPrefix("armeria.server.exceptions.unhandled"),
+                                        MicrometerCounter.class, MicrometerCounter::new);
+        counter = new LongAdder();
+    }
+
+    /**
+     * Increments both {@link #micrometerCounter} and {@link #counter} and stores unhandled exception
+     * that occurs. If multiple servers use the same {@link MeterRegistry}, then {@link #micrometerCounter}
+     * will be shared across servers.
+     */
+    @Nullable
+    @Override
+    public HttpResponse onServiceException(ServiceRequestContext ctx, Throwable cause) {
+        final boolean isExpectedException =
+                (cause instanceof HttpStatusException || cause instanceof HttpResponseException) &&
+                cause.getCause() != null;
```

Comment:
We don't need to calculate this if `ctx.shouldReportUnhandledExceptions()` returns `false`. Please move it inside the `if` block.